core/src/main/java/com/linecorp/armeria/common/logging/RequestScopedMdc.java

```diff
@@ -248,17 +253,28 @@ public static void putAll(RequestContext ctx, Map<String, String> map) {
             return;
         }
 
-        synchronized (ctx) {
-            final Object2ObjectMap<String, String> oldMap = getMap(ctx);
-            final Object2ObjectMap<String, String> newMap;
-            if (oldMap.isEmpty()) {
-                newMap = new Object2ObjectOpenHashMap<>(map);
-            } else {
-                newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
-                newMap.putAll(oldMap);
-                newMap.putAll(map);
+        for (; ; ) {
```

Comment:
I think @ikhoon meant something like the following:

```
         for(;;) {
-            final AttributeKey<ReentrantLock> key = AttributeKey.valueOf("reentrantLock");
-            ReentrantLock oldLock = ctx.attr(key);
-            final ReentrantLock reentrantLock = new ReentrantLock();
-            if (ctx.setAttr(key, reentrantLock) == oldLock) {
-                reentrantLock.lock();
-                try {
-                    final Object2ObjectMap<String, String> oldMap = getMap(ctx);
-                    final Object2ObjectMap<String, String> newMap;
-                    if (oldMap.isEmpty()) {
-                        newMap = new Object2ObjectOpenHashMap<>(map);
-                    } else {
-                        newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
-                        newMap.putAll(oldMap);
-                        newMap.putAll(map);
-                    }
-                    ctx.setAttr(MAP, Object2ObjectMaps.unmodifiable(newMap));
-                } finally {
-                    reentrantLock.unlock();
-                }
+            final Object2ObjectMap<String, String> oldMap = getMap(ctx);
+            final Object2ObjectMap<String, String> newMap;
+            if (oldMap.isEmpty()) {
+                newMap = new Object2ObjectOpenHashMap<>(map);
+            } else {
+                newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
+                newMap.putAll(oldMap);
+                newMap.putAll(map);
+            }
+            if (ctx.setAttr(MAP, Object2ObjectMaps.unmodifiable(newMap)) == oldMap) {
                 break;
             }
         }
``` 

We probably need to modify the other usages of `reentrantLock` in a similar way

Reply:
Since a `setAttr` operation finishes in a very short time, I didn't see we need to use `ReentrantLock` but used a `setAttr` operation and a spin lock.
```java
for (;;) {
    final Object2ObjectMap<String, String> oldMap = getMap(ctx);
    final Object2ObjectMap<String, String> newMap;
    if (oldMap.isEmpty()) {
        newMap = new Object2ObjectOpenHashMap<>(map);
    } else {
        newMap = new Object2ObjectOpenHashMap<>(oldMap.size() + map.size());
        newMap.putAll(oldMap);
        newMap.putAll(map);
    }
    if (ctx.setAttr(MAP, Object2ObjectMaps.unmodifiable(newMap)) == oldMap) {
        break;
    }
}
```

Reply:
Fixed as ikhoon mentioned!

Reply:
Oops, I missed this comment. Let me talk to @ikhoon.